cpp/src/parquet/metadata.cc

```diff
@@ -334,8 +335,17 @@ class ColumnChunkMetaData::ColumnChunkMetaDataImpl {
     return possible_geo_stats_ != nullptr && possible_geo_stats_->is_valid();
   }
 
+  inline std::shared_ptr<EncodedStatistics> encoded_statistics() const {
+    return is_stats_set() ? possible_encoded_stats_ : nullptr;
+  }
+
   inline std::shared_ptr<Statistics> statistics() const {
-    return is_stats_set() ? possible_stats_ : nullptr;
+    if (is_stats_set() && possible_stats_ == nullptr) {
+      // Because we are modifying possible_stats_ in a const method
+      const std::lock_guard<std::mutex> guard(stats_mutex_);
+      possible_stats_ = MakeColumnStats(*column_metadata_, descr_);
+    }
+    return possible_stats_;
```

Comment:
There is a race condition here because of unguarded accesses to `possible_stats_`. You should use something like this instead:
```suggestion
    const std::lock_guard<std::mutex> guard(stats_mutex_);
    if (possible_stats_ == nullptr) {
      possible_stats_ = MakeColumnStats(*column_metadata_, descr_);
    }
    return possible_stats_;
```